### ISSUE TITLE
Interrupting the temporary Eraser tool causes white to be painted instead

### ISSUE THREAD

**Issue by HeCorr** [pencil2d/pencil#1782]
### Issue Summary
When using the Eraser tool via it's temporary shortcut (Ctrl + Shift) and a pop-up window appears (i.e. auto-save or save reminder), the tool is interrupted and the region that was erased turns white as if it was painted.

The same behavior can be observed when Alt-Tabbing.

This does not happen with the actual Eraser tool (shortcut E).

### Actual Results
Erased area turns white.

### Expected Results
Erased area would remain erased.

### Video or Image Reference

https://github.com/pencil2d/pencil/assets/75134774/350e1b73-a0f9-4394-b78b-95290015dfcc

### Steps to reproduce
1. Launch Pencil2D
2. Enable Auto-Save in the preferences and set number of modifications to 4
3. Create a bitmap layer if you don't have one already
4. Paint a large region in dark grey
5. Create a new bitmap layer above the current one and select it
6. Paint a smaller region in black inside the grey region
7. Activate the temporary Eraser tool by holding Ctrl + Shift
8. Erase parts of the black region until the save reminder window appears (may need to release and start a new stroke)
9. Click "No"

### System Information
+ Pencil2D Version:
```
Version: 0.6.6
commit: ac415788dfb21f66e1cbcc1ad60443225a2d0e37
date: 2021-02-17_13:03:37
Qt Version: 5.12.10
```
(Though I believe the issue is still present in the master branch since I've been using a modified version)
+ Operating System:
Windows 10 64-bit

+ RAM Size:
32GB

+ Graphics Tablet:
(irrelevant to this issue, but) Wacom CTL-470

**Comment by HeCorr** [comment]
I believe the issue is due to `EraserTool::pointerReleaseEvent()` calling `Editor::backup()` [here](https://github.com/pencil2d/pencil/blob/160435c6700b6bfc791b8dd70faf027a4f4478c9/core_lib/src/tool/erasertool.cpp#L172) (which ends up triggering an auto-save [here](https://github.com/pencil2d/pencil/blob/160435c6700b6bfc791b8dd70faf027a4f4478c9/core_lib/src/interface/editor.cpp#L324), which opens a temporary popup window) before calling `endStroke()` [here](https://github.com/pencil2d/pencil/blob/160435c6700b6bfc791b8dd70faf027a4f4478c9/core_lib/src/tool/erasertool.cpp#L184).

The ideal fix would be to somehow prevent the temporary eraser from getting messed up when interrupted, but I'm not sure how to do that.

I thought of a few potential solutions:

1. Define a private `bool mSkipSave = false;` variable in [Editor](https://github.com/pencil2d/pencil/blob/160435c6700b6bfc791b8dd70faf027a4f4478c9/core_lib/src/interface/editor.h) and a public `void skipSave();` that sets it to true, call it in `pointerReleaseEvent()` and add a check in `Editor::updateAutoSaveCounter()` like so:
```diff
void Editor::updateAutoSaveCounter()
{
    mAutosaveCounter++;
    if (mAutosaveCounter >= mAutosaveNumber)
    {
-        resetAutoSaveCounter();
-        emit needSave();
+        if (!mSkipSave)
+        {
+            resetAutoSaveCounter();
+            emit needSave();
+        }
+        mSkipSave = false;
    }
}
```

2. Remove the `updateAutoSaveCounter()` call from `Editor::backup()` entirely and update the code to call it directly (probably a bad idea)

3. Somehow postpone the auto-save to after `endStroke()` gets called (no idea how to achieve that though)

Any better ideas?

**Comment by HeCorr** [comment]
I have a potential solution in mind. Can I please be assigned to this?  :)

**Comment by HeCorr** [comment]
So, the best way I can think of to fix this is to prevent temporary tools from being disabled on focus loss (by deleting [this code](https://github.com/pencil2d/pencil/blob/160435c6700b6bfc791b8dd70faf027a4f4478c9/core_lib/src/interface/scribblearea.cpp#L373-L376)), and somehow make Pencil detect whether the keys are no longer being pressed after the focus is regained (otherwise the temporary tool is stuck in the active state until you press and release it's keys once).

Would this be an acceptable fix?

Please note that the temporary tool will not get stuck active if the popup window closes while still holding it's shortcut keys and then release them. It will only happen if the user releases the shortcuts while the popup is active/the main window is out of focus, so maybe that is an acceptable trade-off to fix the Eraser issue.

I can explain my understanding of what's causing the issue soon so you guys come up with a better fix.

**Comment by HeCorr** [comment]
Yeah I figured it would be only event-based.

Maybe we can somehow keep track of key states while the popup is visible, so when it's closed the main window will know whether the temporary keys got released or not if that makes sense.

**Comment by scribblemaniac** [comment]
Keeping track of the key states is not going to be a feasible solution. Qt simply doesn't provide that information, and there are situations, such as alt-tabbing on Windows which will result in the key release event never being called.

I haven't given this issue a whole lot of thought, but I did have a simple idea yesterday. What you could do is split up the functionality of the Editor::backup function into one or more functions to create backup elements, and one function to append a backup element to the history. And the current backup function could be modified to use these functions so that all the calls to backup do not need to be changed. This solves your issue by being able to create a backup element, apply the stroke and then add it to the backup history (which potentially triggers the autosave).

In editor:
```c++
// Backup creation functions
public BackupBitmapElement* createBackupBitmap() { ... }
public BackupVectorElement* createBackupVector() { ... }
...
// Or
public BackupElement* createBackup() { ... }
// Add backup function
public bool addBackup(BackupElement*) { ... }
```
In EraserTool::pointerReleaseEvent:
```c++
{
    BackupElement* backup = mEditor()->createBackup();
    drawStroke();
    mEditor()->addBackup(backup);
```

Having said all that, @J5lx's idea is good too. In fact, I prefer it over my solution, if only because it would not conflict with #1644.

**Comment by HeCorr** [comment]
Thanks for the suggestions! I'll be trying Jakob's idea soon and if that doesn't work I'll try scribblemaniac's.

**Comment by HeCorr** [comment]
Sorry for not working on this yet, I've been busy with other projects.

**Comment by Jose-Moreno** [comment]
@HeCorr I agree with J5lx. This project is a labour of love, and we aren't requiring you to fulfill any quota.

It is already appreciated that you're donating your time and contributing to the software, so take your time and hopefully you can return to this in due time 😃

**Comment by HeCorr** [comment]
Hello. My most sincere apologies but I will no longer be able to work on this. It's been a while since I've even looked at the P2D code and as far as I can remember I was nowhere near to the solution, so please unassign me.

**Comment by chchwy** [comment]
No worries @HeCorr feel free to come back anytime.
